File: packages/expo-conductor/android/src/main/java/expo/modules/conductor/storage/TaskStore.kt
package expo.modules.conductor.storage

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

/**
 * Durable storage for registered tasks, backed by SharedPreferences so that tasks
 * (and the data needed to reschedule alarms after a reboot) survive process death.
 * Mirrors the Web `TaskRegistry` (localStorage) and iOS `TaskStore` (UserDefaults).
 *
 * Every read-modify-write is serialized on a process-wide lock because the WorkManager
 * worker, alarm receiver, FCM service and the JS module all mutate the store from
 * different threads; SharedPreferences `apply()` is last-writer-wins on the whole blob,
 * so without this the compound read→modify→write would lose updates.
 */
class TaskStore(context: Context) {
  private val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

  fun all(): List<JSONObject> = synchronized(LOCK) { readAll() }

  private fun readAll(): List<JSONObject> {
    val raw = prefs.getString(KEY, null) ?: return emptyList()
    val array = JSONArray(raw)
    return (0 until array.length()).map { array.getJSONObject(it) }
  }

  fun get(id: String): JSONObject? = synchronized(LOCK) {
    readAll().firstOrNull { it.optString("id") == id }
  }

  fun upsert(task: JSONObject) = synchronized(LOCK) {
    val list = readAll().filter { it.optString("id") != task.optString("id") }.toMutableList()
    list.add(task)
    persist(list)
  }

  fun remove(id: String): Boolean = synchronized(LOCK) {
    val list = readAll()
    val filtered = list.filter { it.optString("id") != id }
    if (filtered.size == list.size) return@synchronized false
    persist(filtered)
    true
  }

  fun clear() = synchronized(LOCK) { prefs.edit().remove(KEY).apply() }

  /** Whether the conductor is paused. Persisted so it survives process death / reboot. */
  fun isPaused(): Boolean = synchronized(LOCK) { prefs.getBoolean(KEY_PAUSED, false) }

  fun setPaused(value: Boolean) = synchronized(LOCK) { prefs.edit().putBoolean(KEY_PAUSED, value).apply() }

  private fun persist(list: List<JSONObject>) {
    val array = JSONArray()
    list.forEach { array.put(it) }
    prefs.edit().putString(KEY, array.toString()).apply()
  }

  companion object {
    private const val PREFS = "expo.modules.conductor.tasks"
    private const val KEY = "tasks"
    private const val KEY_PAUSED = "paused"
    private val LOCK = Any()
  }
}

File: packages/expo-conductor/ios/Triggers/TaskStore.swift
import Foundation

/// Durable task storage backed by UserDefaults (JSON), mirroring the Android
/// `TaskStore` (SharedPreferences) and Web `TaskRegistry` (localStorage).
///
/// All read-modify-write mutations are serialized on a single static queue because the
/// notification delegate, BGTask handler, push handler and JS module can mutate the store
/// from different threads; without this the compound read→modify→write would lose updates.
final class TaskStore {
  private let defaults = UserDefaults.standard
  private let key = "expo.modules.conductor.tasks"
  private let pausedKey = "expo.modules.conductor.paused"
  private static let queue = DispatchQueue(label: "expo.conductor.taskstore")

  func all() -> [[String: Any]] {
    Self.queue.sync { readAll() }
  }

  private func readAll() -> [[String: Any]] {
    guard let data = defaults.data(forKey: key),
          let array = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]
    else { return [] }
    return array
  }

  func get(_ id: String) -> [String: Any]? {
    Self.queue.sync { readAll().first { ($0["id"] as? String) == id } }
  }

  func upsert(_ task: [String: Any]) {
    Self.queue.sync {
      var list = readAll().filter { ($0["id"] as? String) != (task["id"] as? String) }
      list.append(task)
      persist(list)
    }
  }

  @discardableResult
  func remove(_ id: String) -> Bool {
    Self.queue.sync {
      let list = readAll()
      let filtered = list.filter { ($0["id"] as? String) != id }
      guard filtered.count != list.count else { return false }
      persist(filtered)
      return true
    }
  }

  func clear() { Self.queue.sync { defaults.removeObject(forKey: key) } }

  /// Whether the conductor is paused. Persisted so the headless paths can honor it.
  func isPaused() -> Bool { Self.queue.sync { defaults.bool(forKey: pausedKey) } }

  func setPaused(_ value: Bool) { Self.queue.sync { defaults.set(value, forKey: pausedKey) } }

  private func persist(_ list: [[String: Any]]) {
    guard let data = try? JSONSerialization.data(withJSONObject: list) else { return }
    defaults.set(data, forKey: key)
  }
}
